yaserde/src/lib.rs

```diff
@@ -2,6 +2,78 @@
+//!
+//! YaSerDe makes it easy to serialize XML documents given an properly annotated struct.
+//! Please refer to the `examples` directory for the complete code shown below.
+//!
+//! # Serialize
+//!
+//! For instance, let's say that one wants to generate a XML file for the
+//! [Rust-Embedded community](https://github.com/rust-embedded/). A well known XML
+//! file for microcontrollers is called [SVD](https://github.com/rust-embedded/svd/)
+//! and it can be defined on YaSerDe via structs like so:
+//!
+//!```
+//! use yaserde_derive::YaSerialize;
+//!
+//! #[derive(Default, PartialEq, Debug, YaSerialize)]
+//! #[yaserde(rename = "device")]
+//! struct Device {
+//!   #[yaserde(attribute)]
+//!   schemaversion: String,
+//!   #[yaserde(attribute)]
+//!   xmlns: String,
+//!   #[yaserde(attribute)]
+//!   xsnonamespaceschemalocation: String,
+//!   #[yaserde(child)]
+//!   devattributes: DevAttrs
+//! }
+//! (...)
+//!```
+//!
+//! The interspersed `#[yaserde()]` macros give some indication of what the resulting XML
+//! Will look like, namely, a short snippet of the struct above in XML would be depending on
+//! concrete values passed to the struct (not shown):
+//!
+//!```
+//! <device schemaversion: "1.0-example", xmlns: "ns:.... example"
+//! xsnonamespaceschemalocation: "foo_bar_baz">
+//!    <devattributes>
+//!    </devattributes>
+//! (...)
+//!```
+//!
+//! Notice the important difference in **XML output representation between `attributes` vs
+//! `child`**, since SVD expects information in that particular arrangement. YaSerDe allows that
+//! serialized XML to be valid unlike other Rust XML (de)serialization crates (i.e quick-xml).
+//!
+//! Also the last `DevAttrs` struct field is indeed another struct, so one can chain several
+//! structs to compose the XML structure (again, see examples folder for the complete
+//! example).
+//!
+//! Be mindful that the **Cargo.toml** should not only include `yaserde` and
+//! `yaserde_derive`, but also a few necessary dependencies... [FIXME: THAT FOR SOME
+//! USER-UNFRIENDLY REASON ARE NOT AUTOMATICALLY PULLED IN AS ONE WOULD EXPECT ;P ;P
+//! ... I'm sure there are good reasons, just wanted to leave this like this so that the author
+//! can chip in and comment about the reasons behind that decision.](https://github.com/media-io/yaserde/issues/22) ... **I personally think that issue #22 should be reopened and fixed properly (as in only requiring yaserde as a dependency, adding yaserde_derive as a feature).**
+//!
+//! ```
+//! [dependencies]
+//! # serde = { version = "1.0.123", features = [ "derive" ] }
+//! # quick-xml = { version = "0.21.0", features = [ "serialize" ] }
+//! yaserde = "0.5.1"
+//! yaserde_derive = "0.5.1"
+//! xml-rs = "0.8.3"
+//! log = "0.4"
+//! ```
+//!
+//! Last but not least, in order to have a nice, pretty printed XML output one can do
+//!
+//! ```
+//! PLEASE LET THE USERS KNOW HOW TO DO THAT CLEARLY ON YASERDE???
+//! ```
```

Comment:
very goo start of documentation, for sure it's necessary, I have just not lot of time yet to do it now .... sadly ...
So any contribution is welcome !

Reply:
It's just a note, I suspect that you know already how to prettyprint the XML output?

Reply:
In that it's pretty simple:
Instead of using `yaserde::ser::to_string`, the method: `yaserde::ser::to_string_with_config` allow to pass a [Config](https://docs.rs/yaserde/0.5.1/yaserde/ser/struct.Config.html) object where you can enable `perform_indent`.

Reply:
Exaaactly, I knew I saw it in passing somewhere but couldn't find it again (bad sign). Thanks! I'll add it in the docs now :D

Reply:
Hmm, something like this yields the same 1-line string:

```diff
diff --git a/src/main.rs b/src/main.rs
index f988281..4697cc9 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -179,6 +179,6 @@ fn main() {
                         xsnonamespaceschemalocation: "CMSIS-SVD.xsd".to_string(),
                         devattributes: dev_attrs
                     };
-    
-    println!("{:?}", yaserde::ser::to_string(&dev).ok().unwrap());
+    let yaserde_cfg = yaserde::ser::Config{ perform_indent: true, indent_string: Default::default(), write_document_declaration: Default::default() };
+    println!("{:#?}", yaserde::ser::to_string_with_config(&dev, &yaserde_cfg));
 }
```

How do you print this so that it shows up good? :-S

Reply:
With the changes shown above in the diff:

![Skärmavbild 2021-02-08 kl  22 06 46](https://user-images.githubusercontent.com/175587/107211627-f16cb900-6a59-11eb-91d1-a96b1e1c154b.png)

... still not good :-S

Reply:
Someone has contributed and publish a patch that I have merged few minutes ago, can you rebase from master ?
For the config, you can use the simplest syntax:
```rust
let yaserde_cfg = yaserde::ser::Config{
  perform_indent: true,
  .. Default::default()
};
```

Reply:
I just rebased and pointed to git master on my Cargo.toml, but the prettyprinting is still off:

![Skärmavbild 2021-02-08 kl  22 24 18](https://user-images.githubusercontent.com/175587/107213367-63de9880-6a5c-11eb-9e66-7a380d054f52.png)

Also there's the escaping of the double quotes which also messes up with online XML linters, i.e:

```
<?xml version=\"1.0\" encoding=\"utf-8\"?>
``` 

To which I have to run `cargo` like `cargo run -q | sed -e s'#\\##'g` to get appropriate unquoted output:

![Skärmavbild 2021-02-08 kl  22 27 40](https://user-images.githubusercontent.com/175587/107213638-dd768680-6a5c-11eb-8c6a-f1fdfe90f2e5.png)

```diff
diff --git a/Cargo.toml b/Cargo.toml
index ae31f81..7256f38 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,7 +7,7 @@ edition = "2018"
 [dependencies]
 # serde = { version = "1.0.123", features = [ "derive" ] }
 # quick-xml = { version = "0.21.0", features = [ "serialize" ] }
-yaserde = "0.5.1"
-yaserde_derive = "0.5.1"
+yaserde = { version = "0.5.1", git = "https://github.com/media-io/yaserde/", branch = "master" }
+yaserde_derive = { version = "0.5.1", git = "https://github.com/media-io/yaserde/", branch = "master" }
 xml-rs = "0.8.3"
 log = "0.4"
\ No newline at end of file
diff --git a/src/main.rs b/src/main.rs
index f988281..26a9971 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -180,5 +180,10 @@ fn main() {
                         devattributes: dev_attrs
                     };
     
-    println!("{:?}", yaserde::ser::to_string(&dev).ok().unwrap());
+    // Display pretty printed XML
+    let yaserde_cfg = yaserde::ser::Config{
+        perform_indent: true,
+        .. Default::default()
+    };
+    println!("{:#?}", yaserde::ser::to_string_with_config(&dev, &yaserde_cfg));
 }
\ No newline at end of file
```

Reply:
Sorry I have made a mistake, dependency required to be link with:
```
yaserde = {version = "0.5.1", path = "../yaserde" }
yaserde_derive = {version = "0.5.1", path = "../yaserde_derive" }
```
I have updated your branch to fix it and put this example as test.